### DISCUSSION TITLE
Live Camera View

### DISCUSSION THREAD

**Discussion by SAHorowitz** [SAHorowitz/MSFS2020-PilotPathRecorder#4] [Ideas]
Marty,
  Sorry, it took a while for me to look at this.  I have been out of town unexpectedly and away from my computer.  It is an interesting piece of functionality that you added that I had not thought of for the program prior.  It certainly was a good coding exercise and allowed you to look thru, and understand, the code itself.

I was able to run the code and the new feature functions fine.  There is some lag (maybe 5 seconds?) between the flight in the simulator and Google Earth's reaction. Is that what you saw also?  Also, I noticed that at the beginning of a flight it put the "live view" in the middle of the ocean until you started taxing down the runway.  Did you see that as well?  The base program's logging functionality still works as it did prior to your changes.

I am a bit torn on whether to add this ability to the main program or not since the program is mainly for logging and analysis after the flight is finished.  Do you feel this is an option people will use?  What is the use case for using this feature?

**Comment by noodnik2** [comment]
Thanks for taking a look!

> _some lag (maybe 5 seconds?) between the flight in the simulator and Google Earth's reaction_

The lag shouldn't be much more than the "refresh rate" you configure in the network link.  With a `Time-Based Refresh` of one second, I've found it to be about 1-2 seconds, and with a `View-Based Refresh` of `After Camera Stops` with zero seconds, it's about the same.  It's close enough I've found to get the "feel" of the flight, albeit the scenery is so much worse in Google Earth that it's almost not worth it :slightly_frowning_face:.  But still, it's pretty neat I think to zoom around cities and see the landmarks and other selected features in GE.

> _it put the "live view" in the middle of the ocean_

:nerd_face:  Yea, I see [Null Island](https://en.wikipedia.org/wiki/Null_Island) pretty often, too.  You can see in the code where it calls the `Update` function with the coordinates from `SimConnect` - prior to that call, it'll just take you to (0, 0).  I'm open to suggestions!

> _base program's logging functionality still works as it did_

That was one of the advantages to just storing and using the current coordinates, resulting in very little impact to the other parts of the program.  However, I'm starting now to implement support (on [this branch](https://github.com/noodnik2/MSFS2020-PilotPathRecorder/tree/210119-multitracks)) for [gx:MultiTrack](https://developers.google.com/kml/documentation/kmlreference#gx:multitrack) which will query into the database to render the KML track of all points since the last call to the listener.  Check out the approach I found in an old [Google I/O Talk](https://www.youtube.com/watch?v=nIoWHlEEeNI&t=16m17s) I found on YouTube.  This I think will at least "bridge the gap" somewhat between our sets of functionalities - which until now have been 100% independent of one-another.

I ultimately was looking for the type of "smoothness" that you get from your use of [gx:Tour](https://developers.google.com/kml/documentation/kmlreference#gx:tour), but I couldn't seem to get that working in real-time.  From my experiments, I found that GE just stops refreshing the tour once it's done - no matter what.  Even though there's no `<gx:flyToMode>smooth</gx:flyToMode>` in `gx:MultiTrack`, I think it should result in a better overall update experience, requiring less callbacks.

In any case, yes, I'm getting back into `C#` and this has been an interesting way to do it, giving me an excuse to do some more sim flying.

Regards,

-- Marty

**Reply by SAHorowitz** [reply]
Glad you are having fun with it.  I totally understand the coding and flying -- each gives you an excuse to do the other.

Sounds like you are looking at the multitrack idea.  I did watch the link you sent - very interesting.

Let me know how that goes and we can decide what code to merge up.

Steve

**Comment by noodnik2** [comment]
Hi again Steve,

As you may have seen in my new ["Show and Tell" post](https://github.com/SAHorowitz/MSFS2020-PilotPathRecorder/discussions/9), I finally released [LiveCam Support](https://github.com/noodnik2/MSFS2020-PilotPathRecorder/blob/master/README-kmlcam.md) in my repo.

This side-project gave me the opportunity to learn several things, including:
1. update myself on `C#` and `.NET` (it'd been years!)
2. learn about `KML` and it's ability to help orchestrate a virtual 3D flying experience in Google Earth, and finally;
3. create a `Github Action`, which now builds and publishes a release for me anytime I create and push a tag having a suffix of `-release`

I thought to reply to some points in your previous post:

> _that I had not thought of for the program prior._

For me, that's a really cool part about open source - the fusion of ideas, and organic growth of this intangible thing called "software."

> _some lag (maybe 5 seconds?) between the flight in the simulator and Google Earth's reaction. Is that what you saw also?_

Yea, that's about what I see.  Given the polling going on, and some fundamentals about the underlying design, that seems to be expected.  Please let me know if you've got some ideas about reducing that (short of tweaking down the polling frequency :smirk:).

> _it put the "live view" in the middle of the ocean until you started taxing down the runway._

Thanks for that feedback!  I added some popup dialogs to help manage expectations such as yours if the "LiveCam" isn't ready (e.g., logging is not active, etc.) when the `Link` button is pressed.  I hope this address future embarrassments of this sort :unamused:.  

> _The base program's logging functionality still works as it did prior to your changes._

Thanks for confirming.  I hope that's still the case.  I got confused about that, so I actually wrote a series of unit tests to help convince myself I understood the way that was intended to work.  I suspect you'll find these amusing if you look at them.  :sunglasses:

> _torn on whether to add this ability to the main program or not since the program is mainly for logging and analysis after the flight is finished_

I totally understand.  Other than some of the accompanying changes which you might find useful, at this point I don't hold out an expectation that you'll merge my changes back into your base.  Indeed, my purposes (recounted above), don't seem to be directly in line with your original goals.
